crates/bevy_ecs/src/event/trigger.rs

```diff
@@ -0,0 +1,255 @@
+use crate::{
+    component::ComponentId,
+    entity::Entity,
+    event::{EntityEvent, Event},
+    observer::{CachedObservers, TriggerContext},
+    traversal::Traversal,
+    world::DeferredWorld,
+};
+use bevy_ptr::PtrMut;
+use core::marker::PhantomData;
+
+pub trait Trigger<E: Event> {
```

Comment:
Ah, I see what you mean about this being user-extensible. Yeah okay, I like this!